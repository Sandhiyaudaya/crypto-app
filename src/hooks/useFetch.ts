import { useCallback, useReducer } from 'react';

enum FetchStatus {
  LOADING = 'LOADING',
  LOADED = 'LOADED',
  ERROR = 'ERROR',
}

interface FetchState<T> {
  data?: T;
  error?: string;
  status?: FetchStatus;
}

export interface FetchStateResult<T> extends FetchState<T> {
  isLoading: boolean;
  isError: boolean;
  isSuccess: boolean;
  fetchApi: () => Promise<void>;
}

const initalState = {
  data: undefined,
  status: undefined,
  error: undefined,
};

type Action<T> = { type: FetchStatus; payload?: T };

const reducer =
  <T>() =>
  (state: FetchState<T>, action: Action<T>): FetchState<T> => {
    switch (action.type) {
      case FetchStatus.LOADING:
        return { ...initalState, status: FetchStatus.LOADING };
      case FetchStatus.LOADED:
        return {
          ...initalState,
          status: FetchStatus.LOADED,
          data: action.payload,
        };
      case FetchStatus.ERROR:
        return {
          ...initalState,
          status: FetchStatus.ERROR,
          error: 'Something went wrong. Please try again later',
        };
      default:
        return state;
    }
  };

/*
 Example 
 const { data, error, status, fetchApi,...rest} = useFetch<User>("/users")
 const handleSubmit = async () => {
    await fetchApi();
 }
*/
export const useFetch = <T = unknown>(url: string): FetchStateResult<T> => {
  const [state, dispatch] = useReducer(reducer<T>(), initalState);
  const { status } = state;

  const fetchApi = useCallback(async () => {
    dispatch({ type: FetchStatus.LOADING });
    try {
      const response = await fetch(url);
      const responseData = (await response.json()) as T;
      if (!response.ok || !responseData) {
        throw new Error();
      }
      dispatch({ type: FetchStatus.LOADED, payload: responseData });
    } catch (e) {
      console.error(e);
      dispatch({ type: FetchStatus.ERROR });
    }
  }, [url]);

  return {
    ...state,
    isLoading: status === FetchStatus.LOADING,
    isError: status === FetchStatus.ERROR,
    isSuccess: status === FetchStatus.LOADED,
    fetchApi,
  };
};
