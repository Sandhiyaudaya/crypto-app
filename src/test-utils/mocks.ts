import { vi } from 'vitest';

export const navigationSpy = vi.fn();
export const setSearchParamSpy = vi.fn();

vi.mock('react-router-dom', () => ({
  useSearchParams: () => [{ get: vi.fn() }, setSearchParamSpy],
  useNavigate: () => navigationSpy,
}));

export function createFetchResponse(
  jsonPromise = new Promise((resolve) => resolve({})),
  ok = true,
) {
  return { json: () => jsonPromise, ok };
}

export const MOCK_SINGLE_CURRENCY_DATA = [
  {
    symbol: '1ECO-BTC',
    high: '0.000008000000',
    low: '0.000005770000',
    volume: '645.60697917',
    quoteVolume: '0.00438197',
    percentChange: '-27.00',
    updatedAt: '2023-08-02T09:07:55.697Z',
  },
];

export const MOCK_CURRENCIES_DATA = [
  {
    symbol: '1ECO-BTC',
    high: '0.000008000000',
    low: '0.000005770000',
    volume: '645.60697917',
    quoteVolume: '0.00438197',
    percentChange: '-27.00',
    updatedAt: '2023-08-02T09:07:55.697Z',
  },
  {
    symbol: '3ECO-BTC',
    high: '0.000008000000',
    low: '0.000005770000',
    volume: '645.60697917',
    quoteVolume: '0.00438197',
    percentChange: '-27.00',
    updatedAt: '2023-08-02T09:07:55.697Z',
  },
];
