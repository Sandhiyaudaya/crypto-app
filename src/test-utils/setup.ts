import '@testing-library/jest-dom';
import './mocks';
import { vi } from 'vitest';

beforeAll(() => {
  // Needed for ant Table
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: vi.fn().mockImplementation((query: unknown) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: vi.fn(),
      removeListener: vi.fn(),
      addEventListener: vi.fn(),
      removeEventListener: vi.fn(),
      dispatchEvent: vi.fn(),
    })),
  });
});
