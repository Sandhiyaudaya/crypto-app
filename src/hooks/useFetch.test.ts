import { act, renderHook, waitFor } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';

import { useFetch } from './useFetch';
import { createFetchResponse } from '../test-utils/mocks';

const DEFAULT_DATA = 'Hello';
const DEFAULT_PROMISE = new Promise((resolve) => resolve(DEFAULT_DATA));

describe('useFetch', () => {
  beforeEach(() => {
    global.fetch = vi
      .fn()
      .mockResolvedValue(createFetchResponse(DEFAULT_PROMISE));
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  it('returns error when status is not ok', async () => {
    global.fetch = vi
      .fn()
      .mockResolvedValue(createFetchResponse(DEFAULT_PROMISE, false));
    const { result } = renderHook(() => useFetch('/user'));
    const hookData = result.current;

    await act(async () => {
      await hookData.fetchApi();
    });
    const { isError, error } = result.current;

    expect(isError).toBeTruthy();
    expect(error).toMatchSnapshot();
  });

  it('returns error when data is null', async () => {
    global.fetch = vi
      .fn()
      .mockResolvedValue(
        createFetchResponse(new Promise((resolve) => resolve(null))),
      );
    const { result } = renderHook(() => useFetch('/user'));
    const hookData = result.current;

    await act(async () => {
      await hookData.fetchApi();
    });
    const { isError, error } = result.current;

    expect(isError).toBeTruthy();
    expect(error).toMatchSnapshot();
  });

  it('returns error when promise is rejected', async () => {
    global.fetch = vi
      .fn()
      .mockResolvedValue(
        createFetchResponse(
          new Promise((_resolve, reject) => reject(new Error('API Error'))),
        ),
      );
    const { result } = renderHook(() => useFetch('/user'));
    const hookData = result.current;

    await act(async () => {
      await hookData.fetchApi();
    });
    const { isError, error } = result.current;

    expect(isError).toBeTruthy();
    expect(error).toMatchSnapshot();
  });

  it('returns data when status is ok', async () => {
    const { result } = renderHook(() => useFetch('/user'));
    const { fetchApi } = result.current;

    await act(async () => {
      await fetchApi();
    });
    const { isError, isSuccess, data } = result.current;

    expect(isError).toBeFalsy();
    expect(isSuccess).toBeTruthy();
    expect(data).toBe(DEFAULT_DATA);
  });

  it('returns loading when api is triggered', async () => {
    const { result } = renderHook(() => useFetch('/user'));
    const { fetchApi } = result.current;

    await act(async () => {
      await fetchApi();
    });
    const { isLoading, data, error } = result.current;

    void waitFor(() => {
      expect(isLoading).toBeTruthy();
      expect(data).toBeUndefined();
      expect(error).toBeUndefined();
    });
  });

  it('calls the fetch with the url passed', async () => {
    const url = 'www.myapi/v3/products';
    global.fetch = vi
      .fn()
      .mockResolvedValue(createFetchResponse(DEFAULT_PROMISE, false));
    const { result } = renderHook(() => useFetch(url));

    await act(async () => {
      await result.current.fetchApi();
    });

    expect(global.fetch).toHaveBeenCalledWith(url);
  });
});
