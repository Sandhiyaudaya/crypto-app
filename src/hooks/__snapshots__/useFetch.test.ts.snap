// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useFetch > returns error when data is null 1`] = `"Something went wrong. Please try again later"`;

exports[`useFetch > returns error when promise is rejected 1`] = `"Something went wrong. Please try again later"`;

exports[`useFetch > returns error when status is not ok 1`] = `"Something went wrong. Please try again later"`;
